### PR TITLE
subsys: mgmt: fmfu: Add callback to fmfu init function

### PR DESCRIPTION
**Pull request by sigvartmh**
Add callback so that the user decides what needs to be done before
starting the full modem update process when receiving a full modem
update command on SMP.

The user will have to shutdown the modem and initialize it in DFU_MODE
as described in the documentation of the library. Since we have no
way of checking which mode the library currently is in we can not just
return an error value. As found in ref. NCSDK-9120 having the modem
library not initialized in DFU_MODE will cause the device to reset.

Signed-off-by: Sigvart Hovland <sigvart.hovland@nordicsemi.no>

### CODE REVIEW

**Comment by sigvartmh** [comment]
I Will split it into 2 commits. 

1. For changed that add app update support to the sample
2. Changes to how FMFU interface works

File: subsys/mgmt/fmfu/src/fmfu_mgmt.c [hunk 1]
```diff
@@ -26,6 +26,9 @@ LOG_MODULE_REGISTER(mgmt_fmfu, CONFIG_MGMT_FMFU_LOG_LEVEL);
 
 #define FIRMWARE_SHA_LENGTH 32
 
+static fmfu_callback_t callback;
+static bool should_init = true;
```
Comment (hakonfam) [inline subsys/mgmt/fmfu/src/fmfu_mgmt.c:30]: I think the name of this variable is a bit misleading. To me it sounds like a variable protecting against multiple `init` calls being called (like `initialized` variables typically does).

IMO it would be better to describe what state it represents rather than what action should be performed when it is set.

Just a proposal:

```
static bool update_started; /* Is there an ongoing update */

/* ... */ 

static int fmfu_firmware_upload(struct mgmt_ctxt *ctx)
{
	static bool bootloader = true;
	if (!update_started) {
		callback(FMFU_EVT_INIT);
		update_started = true;
	}
```

Also it would be really good if we could keep this variable inside the scope of the `fmfu_firmware_upload`. This might not be possible, but I would make an effort to see if its possible.

File: samples/nrf9160/fmfu_smp_svr/src/main.c [hunk 2]
```diff
@@ -7,9 +7,41 @@
...
+
+static void fmfu_callback_handler(const enum fmfu_evt_id id)
+{
+	switch (id) {
+	case FMFU_EVT_INIT:
+		/* Shutdown modem to prepare for DFU */
+		nrf_modem_lib_shutdown();
+
+		nrf_modem_lib_init(FULL_DFU_MODE);
+		struct nrf_modem_full_dfu_digest digest;
```
Comment (hakonfam) [inline samples/nrf9160/fmfu_smp_svr/src/main.c:29]: Its most common to declare all variables in the top of the function.

File: samples/nrf9160/fmfu_smp_svr/src/main.c [hunk 3]
```diff
@@ -7,9 +7,41 @@
...
 #include <modem/nrf_modem_lib.h>
 #include <modem/at_cmd.h>
+#ifdef CONFIG_MCUMGR_CMD_IMG_MGMT
+#include "img_mgmt/img_mgmt.h"
+#endif
+#ifdef CONFIG_MCUMGR_CMD_OS_MGMT
+#include "os_mgmt/os_mgmt.h"
+#endif
+
+static void fmfu_callback_handler(const enum fmfu_evt_id id)
```
Comment (hakonfam) [inline samples/nrf9160/fmfu_smp_svr/src/main.c:21]: Is there any point in having `id` const? As its not a pointer I mean. Also it does not seem to match the function signature from the header file.

File: samples/nrf9160/fmfu_smp_svr/src/main.c [hunk 4]
```diff
@@ -7,9 +7,41 @@
...
+
+		int err = nrf_modem_full_dfu_init(&digest);
+
+		if (err != 0) {
+			printk("nrf_fmfu_init failed, err: %d, errno: %d\n.",
+				err, errno);
+		}
+		break;
+	case FMFU_EVT_FINISHED:
+		break;
```
Comment (hakonfam) [inline samples/nrf9160/fmfu_smp_svr/src/main.c:39]: Could we have a `printk` here stating that the update was finished? Just to have something :)

File: include/mgmt/fmfu_mgmt.h [hunk 5]
```diff
@@ -24,15 +24,35 @@
...
+ *
+ */
+typedef void (*fmfu_callback_t)(enum fmfu_evt_id id);
+
 /** @brief Setup and register the command handler for full modem
  *  update command handler group.
  *
- * The modem library must be initialized in DFU mode before calling this
- * function - nrf_modem_lib_init(FULL_DFU_MODE).
+ * The modem library must be initialized in DFU mode when a FMFU_EVT_INIT event
```
Comment (hakonfam) [inline include/mgmt/fmfu_mgmt.h:47]: Shouldn't this be part of the callback documentation? Perhaps list the possible events and state what handling is needed (or only the one event where you do need special handling)

Comment (hakonfam) [reply include/mgmt/fmfu_mgmt.h:47]: Another alternative is to have the docs in the event enum itself

File: samples/nrf9160/fmfu_smp_svr/src/main.c [hunk 6]
```diff
@@ -35,14 +67,16 @@ void main(void)
...
 
 	}
-	/* Shutdown modem to prepare for DFU */
-	nrf_modem_lib_shutdown();
-
-	nrf_modem_lib_init(FULL_DFU_MODE);
 	/* Register SMP Communication stats */
 	fmfu_mgmt_stat_init();
+#ifdef CONFIG_MCUMGR_CMD_OS_MGMT
+	os_mgmt_register_group();
```
Comment (hakonfam) [inline samples/nrf9160/fmfu_smp_svr/src/main.c:73]: How are these relevant to this PR?

File: samples/nrf9160/fmfu_smp_svr/src/main.c [hunk 7]
```diff
@@ -7,9 +7,41 @@
...
+#ifdef CONFIG_MCUMGR_CMD_OS_MGMT
+#include "os_mgmt/os_mgmt.h"
+#endif
+
+static void fmfu_callback_handler(const enum fmfu_evt_id id)
+{
+	switch (id) {
+	case FMFU_EVT_INIT:
+		/* Shutdown modem to prepare for DFU */
+		nrf_modem_lib_shutdown();
```
Comment (hakonfam) [inline samples/nrf9160/fmfu_smp_svr/src/main.c:26]: Could it be an option to have a separate function inside `fmfu_mgmt` which performs these steps instead of requiring the user to invoke these? That way if the API towards `nrf_modem` changes those changes wouldn't be visible for the end user.

```
case FMFU_EVT_INIT:
     /* Do whatever the app needs to do */
     err = fmfu_mgmt_start();
     /* ... */
```

File: samples/nrf9160/fmfu_smp_svr/src/main.c [hunk 8]
```diff
@@ -7,9 +7,41 @@
...
+
+		nrf_modem_lib_init(FULL_DFU_MODE);
+		struct nrf_modem_full_dfu_digest digest;
+
+		int err = nrf_modem_full_dfu_init(&digest);
+
+		if (err != 0) {
+			printk("nrf_fmfu_init failed, err: %d, errno: %d\n.",
```
Comment (hakonfam) [inline samples/nrf9160/fmfu_smp_svr/src/main.c:34]:
```suggestion
			printk("nrf_modem_full_dfu_init failed, err: %d, errno: %d\n.",
```

File: subsys/mgmt/fmfu/src/fmfu_mgmt.c [hunk 9]
```diff
@@ -267,16 +278,13 @@ static struct mgmt_group fmfu_mgmt_group = {
...
-
-	if (err != 0) {
-		LOG_ERR("nrf_fmfu_init failed, errno: %d\n.", errno);
-		return err;
+	if (client_callback == NULL) {
+		return -EINVAL;
 	}
+
+	callback = client_callback;
 	mgmt_register_group(&fmfu_mgmt_group);
```
Comment (hakonfam) [inline subsys/mgmt/fmfu/src/fmfu_mgmt.c:288]: No return value from this function? (I know its not part of this PR, but we could fix it now since we are visiting :D )

File: samples/nrf9160/fmfu_smp_svr/prj.conf [hunk 10]
```diff
@@ -41,3 +41,9 @@ CONFIG_NETWORKING=y
 CONFIG_NET_SOCKETS=y
 CONFIG_NET_SOCKETS_OFFLOAD=y
 CONFIG_HEAP_MEM_POOL_SIZE=8192
+CONFIG_MCUMGR_CMD_IMG_MGMT=y
+CONFIG_MCUMGR_CMD_OS_MGMT=y
+CONFIG_MCUMGR_CMD_STAT_MGMT=y
+
+# Ensure an MCUboot-compatible binary is generated.
```
Comment (hakonfam) [inline samples/nrf9160/fmfu_smp_svr/prj.conf:48]: Should this be here? If so we should include the rationale in the commit message.

File: subsys/mgmt/fmfu/src/fmfu_mgmt.c [hunk 11]
```diff
@@ -267,16 +278,13 @@ static struct mgmt_group fmfu_mgmt_group = {
 	.mg_group_id = (MGMT_GROUP_ID_PERUSER + 1),
 };
 
-int fmfu_mgmt_init(void)
+int fmfu_mgmt_register_group(fmfu_callback_t client_callback)
```
Comment (hakonfam) [inline subsys/mgmt/fmfu/src/fmfu_mgmt.c:281]: IMO we should stick with the `init` name since the user really does not care what happenes in the background here (they don't even know what a "group" is in this context as they are not exposed to it.

**Review by hakonfam — CHANGES_REQUESTED** [review]

File: subsys/mgmt/fmfu/src/fmfu_mgmt.c (hunk 1 above)
Comment (sigvartmh) [reply subsys/mgmt/fmfu/src/fmfu_mgmt.c:30]: sure calling it `update_started` sounds good

File: samples/nrf9160/fmfu_smp_svr/src/main.c (hunk 3 above)
Comment (sigvartmh) [reply samples/nrf9160/fmfu_smp_svr/src/main.c:21]: No point will change it was a pointer in my initial idea.

File: samples/nrf9160/fmfu_smp_svr/src/main.c (hunk 4 above)
Comment (sigvartmh) [reply samples/nrf9160/fmfu_smp_svr/src/main.c:39]: sure

File: include/mgmt/fmfu_mgmt.h (hunk 5 above)
Comment (sigvartmh) [reply include/mgmt/fmfu_mgmt.h:47]: Sure I'll add a bit more to the documentation

File: samples/nrf9160/fmfu_smp_svr/src/main.c (hunk 6 above)
Comment (sigvartmh) [reply samples/nrf9160/fmfu_smp_svr/src/main.c:73]: Supposed to split up in to commits where this is just to add app update to the sample.

File: samples/nrf9160/fmfu_smp_svr/src/main.c (hunk 7 above)
Comment (sigvartmh) [reply samples/nrf9160/fmfu_smp_svr/src/main.c:26]: Had the same idea but then we will hide from the user what is actually happening which is shutting down the modem library. I guess it would be reasonable to assume that they won't use the socket API etc while trying to do this.

Maybe provided a `fmfu_mgmt_default_init()`  function could be an idea so that they know they can fill in what is needed to be done.

File: subsys/mgmt/fmfu/src/fmfu_mgmt.c (hunk 11 above)
Comment (sigvartmh) [reply subsys/mgmt/fmfu/src/fmfu_mgmt.c:281]: the idea here is to unify the API names. Since `os_mgmt_register_group` is a function and `fs_mgmt_register_group` etc. So it would make sense to have the same naming convention across all files registering handlers to MCUMgr

File: subsys/mgmt/fmfu/src/fmfu_mgmt.c (hunk 9 above)
Comment (sigvartmh) [reply subsys/mgmt/fmfu/src/fmfu_mgmt.c:288]: mgmt_register_group does not return any value 
https://github.com/nrfconnect/sdk-mcumgr/blob/b18c7b8e3f1fbfc99e0f02102a9d8e3b60304959/mgmt/include/mgmt/mgmt.h#L367

File: samples/nrf9160/fmfu_smp_svr/src/main.c (hunk 7 above)
Comment (hakonfam) [reply samples/nrf9160/fmfu_smp_svr/src/main.c:26]: Good point, since we state that we need special handling I guess they will take extra care setting this up. Maybe this is something we can describe a bit in the .rst ?

IMO we should either take full ownership of the procedure, or leave the whole procedure to the user.

File: samples/nrf9160/fmfu_smp_svr/src/main.c (hunk 6 above)
Comment (hakonfam) [reply samples/nrf9160/fmfu_smp_svr/src/main.c:73]: Ah, my bad :)

File: samples/nrf9160/fmfu_smp_svr/prj.conf (hunk 10 above)
Comment (sigvartmh) [reply samples/nrf9160/fmfu_smp_svr/prj.conf:48]: Copy pasta will remove
